x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DateEsField.java

```diff
@@ -33,9 +35,15 @@ private DateEsField(StreamInput in) throws IOException {
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
-        out.writeString(getName());
-        out.writeMap(getProperties(), StreamOutput::writeNamedWriteable);
-        out.writeBoolean(isAggregatable());
+        if (((PlanStreamOutput) out).writeEsFieldCacheHeader(this)) {
```

Comment:
One thing that bothers me about the way attributes are cached is that each subclass does this check - and each subclass writes its name *before* checking the cache. It'd be nice if we could flip it. I *think* that'd make `EsField` no longer strictly a `NamedWriteable`. I *guess* what I'd do is put this in the superclass:
```
public final void writeTo(StreamOutput out) {
  PlanStreamOutput pso = (PlanStreamOutput) out;
  if (pso.writeEsFieldCacheHeader(this)) {
    writeString(getWriteableName());
    writeUncached(pso);
  }
}

public abstract void writeUncached(PlanStreamOutput out);
public abstract String getWriteableName();
```

Or something. I'm not really sure exactly how to do it.

Reply:
Uhm... I like it, it will save us a few bytes per field.
Let me give it a try

Reply:
@nik9000 I did it for EsFields, it was a bit convoluted but I think it works.
I can do the same for `FieldAttributes`, but please have a look first and let me know what you think.

Reply:
The only problem with `Attributes` is that they extend `NamedExpression` -> `NamedWriteable`, so I cannot easily make them non-NamedWriteable.
This would make the whole thing terribly error prone, ie. someone could still invoke `writeNamed(attribute)` somewhere, so we'd have to keep both mechanisms in place

Reply:
oh my! That isn't great. Ok. In that case this is probably the right thing to do. Let me look again.

Reply:
It looks like you did it for ESField just fine - but couldn't do it for attributes. Sorry. Didn't read the first comment.